src/cider/common/interpreters/Aggregator.h

```diff
@@ -0,0 +1,211 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+
+#include <type/data/sqltypes.h>
+
+#include <common/Arena.h>
+#include <common/hashtable/FixedHashMap.h>
+#include <common/hashtable/HashMap.h>
+
+namespace Cider {
+
+using AggregateDataPtr = int8_t*;
+
+using AggregatedDataWithUInt8Key =
+    FixedImplicitZeroHashMapWithCalculatedSize<uint8_t, AggregateDataPtr>;
+using AggregatedDataWithUInt16Key = FixedImplicitZeroHashMap<uint16_t, AggregateDataPtr>;
+
+template <typename Base>
+struct AggregationDataWithNullKey : public Base {
+  using Base::Base;
+
+  bool& hasNullKeyData() { return has_null_key_data; }
+  AggregateDataPtr& getNullKeyData() { return null_key_data; }
+  bool hasNullKeyData() const { return has_null_key_data; }
+  const AggregateDataPtr& getNullKeyData() const { return null_key_data; }
+  size_t size() const { return Base::size() + (has_null_key_data ? 1 : 0); }
+  bool empty() const { return Base::empty() && !has_null_key_data; }
+  void clear() {
+    Base::clear();
+    has_null_key_data = false;
+  }
+  void clearAndShrink() {
+    Base::clearAndShrink();
+    has_null_key_data = false;
+  }
+
+ private:
+  bool has_null_key_data = false;
+  AggregateDataPtr null_key_data = nullptr;
+};
+
+template <typename... Types>
+using HashTableWithNullKey = AggregationDataWithNullKey<HashMapTable<Types...>>;
+
+using AggregatedDataWithNullableUInt8Key =
+    AggregationDataWithNullKey<AggregatedDataWithUInt8Key>;
+using AggregatedDataWithNullableUInt16Key =
+    AggregationDataWithNullKey<AggregatedDataWithUInt16Key>;
+
+class Aggregator;
+HashTableAllocator allocator;
+
+struct AggregationMethod : private boost::noncopyable {
+  size_t keys_size{};
+
+  enum class Type {
+    EMPTY = 0,
+    INT8 = 1,
+    INT16 = 2,
+    SERIALIZED = 3,
+    without_key,
+  };
+  Type type = Type::EMPTY;
+
+  AggregationMethod() {}
+
+  ~AggregationMethod();
+};
+
+struct AggKey {
+  bool is_null;
+  int8_t* addr;
+  uint32_t len;
+
+  AggKey(){};
+
+  AggKey(bool is_null, int8_t* addr, uint32_t len) {
+    this->is_null = is_null;
+    this->addr = addr;
+    this->len = len;
+  };
+
+  bool isNull() const { return is_null; }
+  int8_t* getAddr() const { return addr; }
+  uint32_t getLen() const { return len; }
+};
+
+class Aggregator final {

```

Comment:
this is for non-groupby case only? what user should prepare for non-groupby case (like addr, len), from example in ut, addr here is only one row result.

Reply:
It's for group by case only. Users will need a function like `dump` to get all result of the HashTable. The return type of this function has not been decided, maybe it can be CiderBatch, Arrow Array or something else.